axum-core/src/response/into_response_parts.rs

```diff
@@ -116,9 +65,82 @@ where
     V: TryInto<HeaderValue>,
     V::Error: fmt::Display,
 {
-    fn into_response_parts(self, res: &mut ResponseParts) {
+    type Error = TryIntoHeaderError<K::Error, V::Error>;
+
+    fn into_response_parts(self, mut res: ResponseParts) -> Result<ResponseParts, Self::Error> {
         for (key, value) in self {
-            res.insert_header(key, value);
+            let key = key.try_into().map_err(TryIntoHeaderError::key)?;
+            let value = value.try_into().map_err(TryIntoHeaderError::value)?;
+            res.headers_mut().insert(key, value);
```

Comment:
I think this should actually use `append` instead of `insert`, so it's possible to add multiple values for one header.

We should make sure never to set multiple content-types accidentally though.

Reply:
Good point. I'd be a bit worried about always appending. There are probably lots of headers that we should add checks for 🤔

Maybe adding a separate `AppendHeaders` newtype is the safest while keeping insert as the default.

Reply:
I didn't mean there should runtime checks, but rather that the usage patterns in axum and examples should be reviewed.

I think a problem with insert by default is that then inserting via array directly has different behaviour from collecting the array into a HeaderMap and adding that to the response.

I wouldn't expect that to come up in practice usually, but for the eventual case where it comes up, it seems like it could result in a decent amount of confusion and/or serious bugs.

Reply:
Yes thats true! I think `TypedHeader` should probably insert the first value and append any other values. I'll fix that in this PR as well.